src/Features/LanguageServer/Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs

```diff
@@ -136,5 +143,72 @@ void AddDocumentsFromProject(Project? project, ImmutableArray<string> supportedL
+            // diagnostics will be handled by razor itself, which will operate by calling into Roslyn and asking for
+            // document-diagnostics instead.
+            if (context.ClientName != null)
+                return (Documents: ImmutableArray<Document>.Empty, ProjectRental: null);
+
+            using var _ = ArrayBuilder<Document>.GetInstance(out var result);
+
+            if (workspaceDiagnosticsParams?.ProjectId is not null)
+            {
+                var serviceContainer = await _serviceProvider.GetServiceAsync(typeof(SVsBrokeredServiceContainer)).ConfigureAwait(false) as IBrokeredServiceContainer;
+                var serviceBroker = serviceContainer?.GetFullAccessServiceBroker();
+                if (serviceBroker is object)
+                {
+                    var solutionService = await serviceBroker.GetProxyAsync<IEnvironmentService>(FluidServiceDescriptors.EnvironmentService, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    using (solutionService as IDisposable)
+                    {
+                        if (solutionService is object)
+                        {
+                            var projectData = await solutionService.GetProjectByGuidAsync(workspaceDiagnosticsParams.ProjectId.Value, cancellationToken).ConfigureAwait(false);
+                            if (projectData != null)
+                            {
+                                var projectService = await serviceBroker.GetProxyAsync<IProjectService>(projectData.ProjectRequestParameters.GetDescriptor(), projectData.ProjectRequestParameters.Options, cancellationToken).ConfigureAwait(false);
+                                using (projectService as IDisposable)
+                                {
+                                    if (projectService is object)
+                                    {
+                                        var projectRental = await projectService.RequireLanguageServiceAsync(isPrimaryProject: false, cancellationToken).ConfigureAwait(false);
+                                        var project = TryFindLSPProject(_lspWorkspaceRegistrationService, projectData.FullPath);
+                                        if (project is object)
+                                        {
+                                            AddDocumentsFromProject(project, context.SupportedLanguages, isOpen: false, context, result, project.Solution);
+                                            // Ensure that we only process documents once.
+                                            result.RemoveDuplicates();
+                                            return (Documents: result.ToImmutable(), projectRental);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
so as per our offline conversation, i'm of the opinion this woudl/should not be in roslyn.  the idea would be that the host (VS) would ask Snappy to do things like load projects, and would then call into us (using LSP for this feature for example).  From our perspective, we'd not have any idea that anything special was going on.